### PR TITLE
Add Scanner paging support for scan operations to the Cosmos adapter

### PR DESCRIPTION
**Pull request by Torch3333**
This updates the Cosmos Adapter for Scan operation. Up until now, the `Scanner` object, returned by performing a Scan,  downloaded all at once the records eagerly. Now, the Scanner retrieves the records on demand page by page.

### CODE REVIEW

File: core/src/main/java/com/scalar/db/storage/cosmos/SelectStatementHandler.java [hunk 1]
```diff
@@ -89,22 +98,16 @@ private List<Record> executeRead(Get get, TableMetadata tableMetadata) throws Co
                 DSL.field("r.id").eq(cosmosOperation.getId()))
             .getSQL(ParamType.INLINED);
 
-    return getContainer(get).queryItems(query, new CosmosQueryRequestOptions(), Record.class)
-        .stream()
-        .collect(Collectors.toList());
+    return executeQuery(get, tableMetadata, query);
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/storage/cosmos/SelectStatementHandler.java:101]: I'm not sure but should we specify a partition key here?
```suggestion
    return executeQuery(get, tableMetadata, query, new CosmosQueryRequestOptions().setPartitionKey(cosmosOperation.getCosmosPartitionKey()));
```

**Review by brfrn169 — COMMENTED** [review]
Thank you. LGTM overall. Left a question. Please take a look when you have time!

File: core/src/main/java/com/scalar/db/storage/cosmos/SelectStatementHandler.java (hunk 1 above)
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/storage/cosmos/SelectStatementHandler.java:101]: In the sdk [documentation](https://docs.microsoft.com/en-us/azure/cosmos-db/sql/sql-api-query-metrics#about-sql-query-execution) it says 

> When you issue a query to Azure Cosmos DB, the SDK performs these logical steps:
> ...
>     If the query includes a filter against the partition key, like SELECT * FROM c WHERE c.city = "Seattle", **it is routed to a single partition**. If the query does not have a the filter on partition key, then it is executed in all partitions, and results are merged client side.
> ...

So I assume that setting the partition key via the query options is redundant for this query as the where clause already filters on the partition key. This means the query will be routed to a single partition with or without the `CosmosQueryRequestOptions().setPartitionKey` anyway. 

Even though it does not constitute a benchmark, I executed from my laptop to the cosmos integration test machine the query with the added `CosmosQueryRequestOptions().setPartitionKey(cosmosOperation.getCosmosPartitionKey()` and did not notice a significant performance gain or loss compared to the original. A real benchmark would be necessary to see if it improves anything.

Comment (brfrn169) [reply core/src/main/java/com/scalar/db/storage/cosmos/SelectStatementHandler.java:101]: I see. So we don't need to specify a partition key. Thank you for the investigation!

**Review by brfrn169 — APPROVED** [review]
LGTM! Thank you! Left very minor comments. Please take a look when you have time!

**Review by feeblefakie — APPROVED** [review]
LGTM! nice work! Thank you!

File: core/src/main/java/com/scalar/db/storage/cosmos/ScannerImpl.java [hunk 2]
```diff
@@ -14,34 +16,61 @@
...
   @Nonnull
   public Optional<Result> one() {
-    if (records.isEmpty()) {
-      return Optional.empty();
+    while (currentPageRecords.hasNext() || recordsPages.hasNext()) {
+      // Return the next record of the current page if there is one
+      if (currentPageRecords.hasNext()) {
+        Record currentRecord = currentPageRecords.next();
+        Optional<Result> ret = Optional.of(resultInterpreter.interpret(currentRecord));
+        currentPageRecords.remove();
```
[the remark this answers is not in the document]
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/storage/cosmos/ScannerImpl.java:47]: No, I did not think this through enough. I will remove it.

File: core/src/main/java/com/scalar/db/storage/cosmos/ScannerImpl.java [hunk 3]
```diff
@@ -14,34 +16,61 @@
...
+        return ret;
+        // Otherwise, advance to the next page
+      } else {
```
[the remark this answers is not in the document]
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/storage/cosmos/ScannerImpl.java:50]: Yes!

**Comment by Torch3333** [comment]
It appears the PR was merged but I still have a few changes to apply so I will do it in another PR.

**Pull request merged**